### ISSUE TITLE
goProbe rotation runs into deadlock

### ISSUE THREAD

**Issue by fako1024** [els0r/goProbe#71]
When running a current instance of goProbe (`develop`) I can see flows sporadically not being tracked properly. Compared to a legacy version of goProbe there are several missing (although the overall numbers are not crazily far off, so it must be an issue that does not affect _all_ flows). Both queries were issued at 16:00, so the flows should have long rotated for sure:
```
# goQuery_legacy -d /usr/local/goProbe/db -i eth1.10 -f "2023-03-07 15:10:00" -c "dip = 84.46.64.49" -n 10 raw

                                                                      packets   packets             bytes      bytes       
               time    iface         sip          dip  dport  proto        in       out      %         in        out      %
  23-03-07 15:23:49  eth1.10  10.0.0.102  84.46.64.49    443    TCP    2.37 k    5.63 k   4.15  155.81 kB    7.84 MB   4.03
  23-03-07 15:38:49  eth1.10  10.0.0.102  84.46.64.49    443    TCP  488.00      1.58 k   1.07   32.62 kB    2.18 MB   1.12
  23-03-07 15:48:49  eth1.10  10.0.0.102  84.46.64.49    443    TCP   49.94 k  132.69 k  94.77    3.17 MB  184.87 MB  94.85
  23-03-07 15:53:49  eth1.10  10.0.0.102  84.46.64.49    443    TCP    4.00      8.00     0.01  264.00  B  576.00  B   0.00
                                                                                                                           
                                                                      52.81 k  139.92 k           3.35 MB  194.89 MB       
                                                                                                                    
            Totals:                                                            192.72 k                    198.25 MB  

Timespan / Interface : [2023-03-07 15:07:45, 2023-03-07 15:58:49] / eth1.10
Sorted by            : first packet time
Query stats          : 4.00   hits in 89ms
Conditions:          : dip = 84.46.64.49
```
vs.
```
# goQuery_slimcap -d /tmp/test_db -i eth1.10 -f "2023-03-07 15:10:00" -c "dip = 84.46.64.49" -n 10 raw

                                                                       packets  packets              bytes    bytes        
                 time    iface         sip          dip  dport  proto       in      out       %         in      out       %
  2023-03-07 15:25:00  eth1.10  10.0.0.102  84.46.64.49    443    TCP   2.37 k   5.63 k  100.00  155.81 kB  7.84 MB  100.00
                                                                                                                           
                                                                        2.37 k   5.63 k          155.81 kB  7.84 MB        
                                                                                                                   
              Totals:                                                            8.01 k                     7.99 MB  

Timespan / Interface : [2023-03-07 15:10:00, 2023-03-07 15:30:00] / eth1.10
Sorted by            : first packet time
Query stats          : displayed top 1 hits out of 1 in 12ms
Conditions:          : dip = 84.46.64.49
```

**Comment by fako1024** [comment]
@els0r FYI. Not sure yet what's going on - I'll try to dig down and figure it out...

**Comment by fako1024** [comment]
Doesn't seem to be a matter of losing packets, just not taking them into account (correctly)...:
```
2023-03-07T16:35:00.086+0100    info    capture/capture.go:195  Got stats on eth1.10: capture.CaptureStats{PacketsReceived:101227, PacketsDropped:0}    {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.10", "state": "capturing"}
```

**Comment by fako1024** [comment]
@els0r I'm afraid this looks like another state machine issue (maybe even related to #66 ). Both legacy / current versions were started at the same time (and kept running of course), but apparently the current version only has _one_ block written, while the legacy version has rotated more often:
```
# ll /tmp/db_legacy/eth1.10/1678147200/
total 140
drwxr-xr-x 2 root root  4096 Mar  7 16:50 .
drwxr-xr-x 3 root root  4096 Mar  7 16:50 ..
-rw-r--r-- 1 root root 14791 Mar  7 17:05 bytes_rcvd.gpf
-rw-r--r-- 1 root root 14880 Mar  7 17:05 bytes_sent.gpf
...
```
```
# ll /tmp/test_db/eth1.10/1678147200/
total 44
drwxr-xr-x 2 root root 4096 Mar  7 16:50 .
drwxr-xr-x 3 root root 4096 Mar  7 16:50 ..
-rw------- 1 root root  232 Mar  7 16:50 .blockmeta
-rw-r--r-- 1 root root  427 Mar  7 16:50 bytes_rcvd.gpf
-rw-r--r-- 1 root root  569 Mar  7 16:50 bytes_sent.gpf
```
```
              Column: bytes_rcvd
    Number of Blocks: 1
     Number of Flows: 277 / 310 (IP v4 / v6)
             Traffic: bytes: received=81372891 sent=11152238; packets: received=83281 sent=41226
                Size: 1734 bytes

        #            ts               time UTC    offset     len    raw len    encoder     ratio    #F v4    #F v6    drops    first 4 bytes        integrity check
                                                                                                                                                                       
        1    1678204200    2023-03-07 15:50:00         0    1734       2349        lz4    73.82%      277      310        0         04224d18 
```

You can also see from the logs that there was only one rotation being performed (it was 17:13 at the time I copied the following lines):
```
2023-03-07T16:45:48.256+0100    info    writeout/handler.go:132 waiting for 4m12s to start capture rotation     {"app_name": "goProbe_slimcap", "app_version": "872abf81"}
2023-03-07T16:45:48.260+0100    info    capture/capture.go:324  initializing capture    {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1", "state": "initializing"}
2023-03-07T16:45:48.255+0100    info    writeout/handler.go:180 starting writeout handler       {"app_name": "goProbe_slimcap", "app_version": "872abf81"}
2023-03-07T16:45:48.262+0100    info    capture/capture.go:324  initializing capture    {"app_name": "goProbe_slimcap", "app_version": "872abf81", "state": "initializing", "iface": "eth1.1"}
2023-03-07T16:45:48.301+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.11", "state": "capturing"}
2023-03-07T16:45:48.301+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.16", "state": "capturing"}
2023-03-07T16:45:48.308+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.10", "state": "capturing"}
2023-03-07T16:45:48.312+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "t_c1_fw1", "state": "capturing"}
2023-03-07T16:45:48.316+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.1", "state": "capturing"}
2023-03-07T16:45:48.321+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.15", "state": "capturing"}
2023-03-07T16:45:48.322+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "t_c1_fw2", "state": "capturing"}
2023-03-07T16:45:48.329+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "tun2", "state": "capturing"}
2023-03-07T16:45:48.328+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1", "state": "capturing"}
2023-03-07T16:45:48.330+0100    info    capture/capture.go:344  capturing packets       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "ppp0", "state": "capturing"}
2023-03-07T16:50:00.128+0100    info    capture/capture.go:195  Got stats on eth1.10: capture.CaptureStats{PacketsReceived:87456, PacketsDropped:0}     {"app_name": "goProbe_slimcap", "app_version": "872abf81", "state": "capturing", "iface": "eth1.10"}
2023-03-07T16:50:00.128+0100    info    capture/capture.go:195  Got stats on eth1: capture.CaptureStats{PacketsReceived:111900, PacketsDropped:0}       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1", "state": "capturing"}
2023-03-07T16:50:00.130+0100    info    capture/capture.go:195  Got stats on ppp0: capture.CaptureStats{PacketsReceived:124524, PacketsDropped:0}       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "ppp0", "state": "capturing"}
2023-03-07T16:50:00.644+0100    info    capture/capture.go:195  Got stats on t_c1_fw1: capture.CaptureStats{PacketsReceived:4978, PacketsDropped:0}     {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "t_c1_fw1", "state": "capturing"}
2023-03-07T16:50:00.664+0100    info    capture/capture.go:195  Got stats on t_c1_fw2: capture.CaptureStats{PacketsReceived:2831, PacketsDropped:0}     {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "t_c1_fw2", "state": "capturing"}
2023-03-07T16:50:01.256+0100    info    capture/capture.go:195  Got stats on eth1.11: capture.CaptureStats{PacketsReceived:23902, PacketsDropped:0}     {"app_name": "goProbe_slimcap", "app_version": "872abf81", "state": "capturing", "iface": "eth1.11"}
2023-03-07T16:50:05.672+0100    info    capture/capture.go:195  Got stats on eth1.16: capture.CaptureStats{PacketsReceived:592, PacketsDropped:0}       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.16", "state": "capturing"}
```

**Comment by fako1024** [comment]
Note that I added the `Got stats on` in `capture.go`;
```
...
		stats := c.tryGetCaptureStats()
		lastRotationStats := *stats

		logger.Infof("Got stats on %s: %#v", c.iface, *stats)
...
```

**Comment by fako1024** [comment]
Fun fact: After quite a while (25min) after the last one it now _did_ rotate (no interaction on my part):
```
2023-03-07T17:14:29.004+0100    info    capture/capture.go:195  Got stats on eth1.1: capture.CaptureStats{PacketsReceived:6, PacketsDropped:0}  {"app_name": "goProbe_slimcap", "app_version": "872abf81", "state": "capturing", "iface": "eth1.1"}
2023-03-07T17:14:29.085+0100    info    capture/capture.go:195  Got stats on eth1: capture.CaptureStats{PacketsReceived:552652, PacketsDropped:0}       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1", "state": "capturing"}
2023-03-07T17:14:29.091+0100    info    capture/capture.go:195  Got stats on ppp0: capture.CaptureStats{PacketsReceived:615806, PacketsDropped:0}       {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "ppp0", "state": "capturing"}
2023-03-07T17:14:29.100+0100    info    capture/capture.go:195  Got stats on t_c1_fw2: capture.CaptureStats{PacketsReceived:16334, PacketsDropped:0}    {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "t_c1_fw2", "state": "capturing"}
2023-03-07T17:14:29.141+0100    info    capture/capture.go:195  Got stats on eth1.10: capture.CaptureStats{PacketsReceived:491493, PacketsDropped:0}    {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.10", "state": "capturing"}
2023-03-07T17:14:30.108+0100    info    capture/capture.go:195  Got stats on t_c1_fw1: capture.CaptureStats{PacketsReceived:30044, PacketsDropped:0}    {"app_name": "goProbe_slimcap", "app_version": "872abf81", "state": "capturing", "iface": "t_c1_fw1"}
2023-03-07T17:14:33.056+0100    info    capture/capture.go:195  Got stats on eth1.16: capture.CaptureStats{PacketsReceived:3212, PacketsDropped:0}      {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.16", "state": "capturing"}
2023-03-07T17:15:47.720+0100    info    capture/capture.go:195  Got stats on eth1.1: capture.CaptureStats{PacketsReceived:2, PacketsDropped:0}  {"app_name": "goProbe_slimcap", "app_version": "872abf81", "iface": "eth1.1", "state": "capturing"}
```

**Comment by fako1024** [comment]
@els0r Yeah, pretty straightforward. This select is waiting for the rotation to complete (even if not hit during a rotation, e.g. when the capture is waiting / blocking:
```
// map access needs to be synchronized. This will incur a performance hit, but less so
		// than locking the data structure itself
		select {
		case <-c.inRotation:     /// <==== HERE
			select {
			case <-c.rotationDone:
				continue
			case <-c.ctx.Done():
				logger.Debug("context cancelled during rotation. Exiting")
				return
			}
```
Since this is a blocking channel, the `inRotation` _must_ be consumed in the select, although we just want to wait _if_ we happen to be in rotation when the next packet comes in. Wouldn't your idea of an atomic bool fit this scenario much more nicely? :thinking:

**Comment by fako1024** [comment]
Duplicate of #66 .

**Comment by fako1024** [comment]
In order to assess the impact of different ways of ensuring synchronicity I've implemented a Benchmark with a Mock capture source for the whole `process()` routine (unfortunately I had to jump through hoops because since it's an infinite for / select statement I can't use the normal benchmark for loop). On my machine the whole `process()` iteration takes about (37 ± 1) ns using our current (slightly racy) logic.
When switching that to a `sync.Mutex` (which uses an atomic counter already under the hood for the "golden path" we're looking at, so probably is the best we can achieve) this increases to about (57 ± 1) ns, so about 50% higher.

However, there may be a way to achieve a non-racy rotation using an explicit event in the capture routine (to unblock the `PPOLL`) and by explicitly confirming the reception of the attempt to rotate from `process()`. I'm currently testing that, initial attempts look good. I'll draft a PR as soon as I have confirmation...

**Comment by fako1024** [comment]
Restructured the whole rotation logic by creating a three-stage channel-based lock mechanism in combination with an `Unblock()` event handler for the slimcap capture event file descriptor (allowing to manually break the loop when required without having to rely on permanent timeout / retry logic). Since it does not change the `select` logic during normal packet processing there is zero performance impact (or close to it, since there are a few minor checks that have to be done in slimcap, but they only boil down to a few non-atomic boolean checks per second):
```
=== RUN   TestMockPacketCapturePerformance
Packets processed after 10s: 275307275 (36ns/pkt)
```

I've run the following tests to ensure there is no remaining race condition:

- Manually reproduced the issue and used that as testbed for finding a / the solution
- Added and ran tests that simulate the whole `process()` / `rotate()` logic using a mock `captureHandler` with the race detector enabled (which found the race prior to this change basically immediately)
- Ran a multi-interface goProbe (high, low and zero traffic interfaces included) overnight with rotation interval set to once per second

All checks pan out now, so I'm relatively confident that this is sane. PR will follow in a moment.

**Comment by fako1024** [comment]
I'm afraid I have to reopen this. While implementing #88 and using an _actual_ mock source the issue resurfaced (now quite reliably failing on one of the `Deadlock` tests, especially with high (simulated) traffic). I'll investigate (again)...

**Comment by fako1024** [comment]
For low traffic situations it seem the issue is not the actual rotation, but the `Close()` operation. Although it passes successfully, the (mock) source is still waiting / blocking:
```
goroutine 156 [syscall]:
syscall.Syscall(0x493d63?, 0xc0000720c0?, 0xc00012eea0?, 0x442645?)
	/usr/local/go/src/syscall/syscall_linux.go:69 +0x27 fp=0xc000059a68 sp=0xc0000599f8 pc=0x478807
syscall.Syscall(0x0, 0x6, 0xc000059b08, 0x8)
	<autogenerated>:1 +0x34 fp=0xc000059a98 sp=0xc000059a68 pc=0x478db4
github.com/fako1024/slimcap/capture/afpacket/socket.read(0x5b80fb?, {0xc000059b08?, 0xc000059b20?, 0x2?})
	/home/fako/Develop/go/src/github.com/fako1024/slimcap/capture/afpacket/socket/socket_mock.go:105 +0x3f fp=0xc000059ae8 sp=0xc000059a98 pc=0x5b77bf
github.com/fako1024/slimcap/capture/afpacket/socket.(*MockFileDescriptor).ReleaseSemaphore(0x10000003a?)
	/home/fako/Develop/go/src/github.com/fako1024/slimcap/capture/afpacket/socket/socket_mock.go:84 +0x38 fp=0xc000059b20 sp=0xc000059ae8 pc=0x5b7718
github.com/fako1024/slimcap/event.(*Handler).Poll(0xc0000ec078, 0xa0ae?)
	/home/fako/Develop/go/src/github.com/fako1024/slimcap/event/poll.go:47 +0x65 fp=0xc000059b50 sp=0xc000059b20 pc=0x5b7ce5
github.com/fako1024/slimcap/capture/afpacket/afring.(*Source).nextPacket(0xc000138120)
	/home/fako/Develop/go/src/github.com/fako1024/slimcap/capture/afpacket/afring/afring.go:351 +0x685 fp=0xc000059d10 sp=0xc000059b50 pc=0x5b8ca5
github.com/fako1024/slimcap/capture/afpacket/afring.(*Source).NextIPPacketZeroCopy(0xc000138120)
	/home/fako/Develop/go/src/github.com/fako1024/slimcap/capture/afpacket/afring/afring.go:241 +0x25 fp=0xc000059d30 sp=0xc000059d10 pc=0x5b8165
github.com/fako1024/slimcap/capture/afpacket/afring.(*MockSource).NextIPPacketZeroCopy(0xc0000141c0?)
	<autogenerated>:1 +0x25 fp=0xc000059d48 sp=0xc000059d30 pc=0x5ba7e5
github.com/els0r/goProbe/pkg/capture.(*Capture).capturePacket(0xc000118be0)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/capture/capture.go:426 +0x3f fp=0xc000059e08 sp=0xc000059d48 pc=0x5f565f
github.com/els0r/goProbe/pkg/capture.(*Capture).process(0xc000118be0)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/capture/capture.go:491 +0x86 fp=0xc000059ea0 sp=0xc000059e08 pc=0x5f58c6
github.com/els0r/goProbe/pkg/capture.testDeadlock(0xc0005d6000, 0x0)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/capture/capture_test.go:99 +0x26f fp=0xc000059f48 sp=0xc000059ea0 pc=0x5f874f
github.com/els0r/goProbe/pkg/capture.TestLowTrafficDeadlock(0x0?)
	/home/fako/Develop/go/src/github.com/els0r/goProbe/pkg/capture/capture_test.go:18 +0x25 fp=0xc000059f70 sp=0xc000059f48 pc=0x5f7f65
testing.tRunner(0xc0005d6000, 0x694d28)
	/usr/local/go/src/testing/testing.go:1576 +0x10b fp=0xc000059fc0 sp=0xc000059f70 pc=0x4e1c0b
testing.(*T).Run.func1()
	/usr/local/go/src/testing/testing.go:1629 +0x2a fp=0xc000059fe0 sp=0xc000059fc0 pc=0x4e2c4a
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1598 +0x1 fp=0xc000059fe8 sp=0xc000059fe0 pc=0x46c281
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1629 +0x3ea
```
This might hence be an artifact of the mock source not handling the close operation in the same manner (but it could also be a symptom that would manifest on a _real_ source).

**Comment by fako1024** [comment]
Issues are fixed in https://github.com/fako1024/slimcap/issues/22 , will bump `slimcap` dependency in #88 .
